CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out

veronikaslc opened the pull request:
**Overview of changes visible to the user:**
* A new questionnaire property labeled **Require completion** has been added and can be updated in the questionnaire designer
* The PREMs surveys (CPESIC, OAIP, OED, Rehab) have been updated to have this property set to true
* When a form requires completion & the user saves or clicks next in the pagination, the page scrolls to the 1st incomplete question on the screen which gets outlined in red

**Testing**
* Check that nothing changed when filling out a form that does NOT require completion, both as a normal user and as a patient
* Fill in forms with required completion and check that the behavior matches the description above
* Fill in the PREMs surveys as a patient and check that you can't advance to the next page not to the next survey if all the questions on the current screen aren't complete

**Regression testing**
Test that the code refactoring didn't affect the following:
* The color of the answer instructions (e.g. "This question is mandatory") correctly reflects the presence of the answer in all questions including QuestionMatrix
* Incomplete questions and sections appear in view mode, while non-mandatory questions without an answer are not displayed in view mode

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Outlne the incomplete mandatory question with error color instead of primary color

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Better prop name for disabling a main action button

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Re-fetch the form after save to see if user can progress ONLY if the form is required to be complete

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Revert hard-coded completion requirement from the QuestionnaireSet component. It will be replaced by a configuration.

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
* Rename `requiteComplete` to `requireCompletion`
* Add `requireCompletion` as a questionnaire property that can be overridden by the
  `requireCompletion` prop of the `Form` component
* Simplify some conditions by removing unnecessary checks of the completion requirement
  when the condition is also depending on incompleteQuestionEl (as incompleteQuestionEl
  can only be defined if completion is required)
* Minor improvements to code comments

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Set `requireCompletion` to True for PREMs surveys

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -235,8 +238,21 @@ function Form (props) {
         setLastSaveStatus(undefined);
-    })
-    .finally(() => {formNode?.current && setSaveInProgress(false)});
+    }).finally(() => {
+        // Re-fetch the form after save to see if user can progress
+        fetchWithReLogin(globalLoginDisplay, formURL + '.deep.json')
+          .then((response) => response.ok ? response.json() : Promise.reject(response))
+          .then(json => {
+            let incompleteEl = getFirstIncompleteQuestionEl(json);
+            setIncompleteQuestionEl(incompleteEl);
+            !incompleteEl && onSuccess?.();
+            setLastSaveStatus(true);
+            setLastSaveTimestamp(new Date());
+          })
+          .catch(handleFetchError)
+          .finally(() => {
+            formNode?.current && setSaveInProgress(false)});
+          });
```
We should only re-fetch if `requireComplete` is true

marta- replied:
For very long paginated forms this re-fetch of the whole form json is very wasteful. Can we fetch only the current page?

marta- replied:
> We should only re-fetch if `requireComplete` is true

Done in 4e92370.

> Can we fetch only the current page?

This has not yet been addressed. Not a blocker, but a nice to have.

veronikaslc replied:
By thy `current page` do you mean just the current the questionnaire out of the lisr that user is progressing through?

veronikaslc replied:
I think (when paginating) we could just fetch the path of the question that blocked the advancing and if completed - advance further.

marta- replied:
Indeed. I'd like this done in this iteration, but if it takes too long we can do it later as an improvement.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx:
```diff
@@ -213,7 +215,7 @@ function QuestionnaireSet(props) {
   useEffect(() => {
     if (!subjectData || !questionnaireIds) return;
-    setComplete(Object.keys(subjectData || {}).filter(q => isFormComplete(q)).length == questionnaireIds.length);
+    setComplete(questionnaireIds.filter(q => isFormComplete(q)).length == questionnaireIds.length);
```
Optimizations that are unrelated to the main task should be in separate [misc] commits

marta- replied:
Removed when the changes to QuestionnaireSet were reverted in favor of a configuration in f80977b. @veronikaslc please make a separate [misc] PR later with this optimization.

marta- reviewed (dismissed):
**The reference questions are again causing an "AnswerSections out of order" bug.**

Steps to replicate:
* create a visit for the ED clinic
* after the ED survey is automatically generated, deleted and create a new one manually to ensure the reference answers are actually filled in
* open the ED survey as admin (no need to go to the patient portal, the behavior will be the same)
* fill in the 1st question and click NEXT -> **you advance to the next page and the first question there is highlighted: wrong**
* to test non-paginated surveys go to the questionnaire designer > ED survey and set Paginate to false
* go back to editing the ED form (now all on one page), fill in the second question anch click save -> **the last question is the one highlighted instead of the 3rd one - wrong**

This is happening because the AnswerSections are not in the same order as the sections from the questionnaire definition due to autocreation of reference answers. Suggested fix: when finding the first unanswered question, follow the order of the questionnaire instead of the one in the form.

The fact that we **advance to the next page** in the paginated form even though the current one is incomplete appears to be a **separate bug**.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -213,15 +230,12 @@ function Form (props) {
         return;
       }
       if (response.ok) {
-        setLastSaveStatus(true);
-        setLastSaveTimestamp(new Date());
```
These should not be moved down to lines 261-262. We did save, the data is on the server, we're simply not allowing the user to move forward if a complete form is required and they still have missing answers.

marta- replied:
Change reverted in c10ab16.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -213,15 +230,12 @@ function Form (props) {
             .then(json => setStatusFlags(json.statusFlags))
             .catch(err => console.log("The form status flags could not be updated after saving"));
         }
-        onSuccess?.();
```
Should this be `!requireCompletion && onSuccess?.()`?

marta- replied:
Actually we need to keep onSuccess regardless of completion status, since it refers strictly to the success of sending the data to the backend. In practice, it appears to be used for removing listeners when closing the window.

marta- replied:
Change reverted in c10ab16.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -235,8 +249,24 @@ function Form (props) {
         setLastSaveStatus(undefined);
-    })
-    .finally(() => {formNode?.current && setSaveInProgress(false)});
+    }).finally(() => {
+        // If the form is required to be complete, re-fetch it after save to see if user can progress
+        if (requireCompletion) {
+            fetchWithReLogin(globalLoginDisplay, formURL + '.deep.json')
+              .then((response) => response.ok ? response.json() : Promise.reject(response))
+              .then(json => {
+                  let incompleteEl = getFirstIncompleteQuestionEl(json);
+                  setIncompleteQuestionEl(incompleteEl);
+                  !incompleteEl && onSuccess?.();
+                  setLastSaveStatus(true);
+                  setLastSaveTimestamp(new Date());
+              })
+              .catch(handleFetchError)
+              .finally(() => {formNode?.current && setSaveInProgress(false)});
+        } else {
+            formNode?.current && setSaveInProgress(false);
+        }
+    });
```
Why are we re-fetching in `finally` instead of on the success branch?

marta- replied:
Moved in c10ab16.

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
* Pass proper boolean values to the boolean props for disabling the save buttons
* Also listen to changes of the `disableProgress` prop in pagination when launching the next page effect

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
* Revert changes: always mark the save as successfull if the data made it to the backend, regardless of completion status
* Only re-fetch the form data to check completion if the save was successful

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Looping through the questions in a right order
Reseting params values to the original on page change and on the save
events

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Code optimisation not to rentderpagination in view mode
Better codestyle

marta- commented:
**Bug**: For non-paginated forms that require completion, the save button is always disabled.

Steps to replicate:
* In the questionnaire designer, make a questionnaire with Paginate = False and Require completion = true (for example, for simplicity turn on require completion on Patient information, or turn off pagination in the ED survey in `prems` mode)
* Create a form- for that questionnaire - note how the save button is disabled at the beginning (change in behavior, but no big deal yet)
* Start filling out answers - **the save button remains disabled - wrong**
* Fill out everything (mandatory or not) - **the save button is still disabled - wrong**
* Note: clicking on the "Save & view" checkmark at the top or simply leaving the page results in the data being correctly saved as expected

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Fixed non-paginated form save button disabled

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -162,7 +165,6 @@ function FormPagination (props) {
       startIcon={activePage === lastValidPage() ? doneIcon : undefined}
       type="submit"
       variant="contained"
-      disabled={saveInProgress}
```
Why was this line removed?

veronikaslc replied:
Not sure, probably an accidental thing

veronikaslc replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1282/commits/1fb7f563493d1941bde39e9ee63e1c9395a8ed09

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Added accidenltly removed line

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -185,6 +211,10 @@ function Form (props) {
     setError(response);
     setData([]);  // Prevent an infinite loop if data was not set
+    if (isEdit) {
+      paginationEnabled && setDisableProgress(requireCompletion);
```
Not sure we want to disable progress on fetch error, can you motivate the decision?

veronikaslc replied:
Just resetting all of the params to the default values. That is not necessary disabling - `requireCompletion` could be false.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -409,8 +454,90 @@ function Form (props) {
             </div>
   )
 
+  // A helper recursive function to loop through the sections/questions data 
+  // and collect all questions that have incomplete status among it's statusFlags
+  // Recursively build a map of uuid->path of all incomplete questions
+  let getIncompleteQuestionsMap = (sectionJson) => {
```
This Form component is really bloated right now. Wdyt about moving these functions in a separate utility file?

veronikaslc replied:
Totally agree

veronikaslc replied:
Addressed in [30f765e](https://github.com/data-team-uhn/cards/pull/1282/commits/30f765ef97e60db2795d3e0fb62f33f284300b72)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -508,13 +635,15 @@ function Form (props) {
               setPagesCallback={setPages}
-              onDone={() => { setEndReached(true) }}
+              onDone={() => { setEndReached(true); }}
+              onPageChange={() => { setDisableProgress(requireCompletion); setIncompleteQuestionEl(null); }}
```
I'm not sure I understand why in paginated forms progress is disabled at first on each page, can you elaborate?

veronikaslc replied:
that is a long story. In short - `FormPagination` component has `useEffect()` that watches the combinations of 3 parameters:
`saveInProgress, pendingSubmission, disableProgress`
There are certain moments when we click on `Next` button to progress, but save did not just start yet. If we enable progress but default at all times - we immediately go to the next page disregarding all the necessary incomplete checks.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -169,8 +190,13 @@ function Form (props) {
-    setPaginationEnabled(!!json?.['questionnaire']?.['paginate'] && isEdit);
+    
     if (isEdit) {
+	  setPaginationEnabled(!!json?.['questionnaire']?.['paginate']);
```
Tiny indentation mismatch here

```suggestion
      setPaginationEnabled(!!json?.['questionnaire']?.['paginate']);
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -144,6 +159,12 @@ function Form (props) {
   }, [isEdit]);
+  
+  useEffect(() => {
+    // If reuired complition is set, stop any advancing progress untill check that all required 
+    // questions are completed
```
```suggestion
    // If `requireCompletion` is set, stop any advancing progress until check that all required 
    // questions are completed
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -112,8 +119,16 @@ function Form (props) {
-    lastSaveStatus && endReached && onDone && onDone();
-  }, [lastSaveStatus, endReached]);
+    // If there's at least a question that is incomplete while we require completion,
+    // focus on that element and fo not call `onDone`
```
```suggestion
    // focus on that element and do not call `onDone`
```

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out

Misc typo and indentation fixes

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -409,8 +454,90 @@ function Form (props) {
+    let retFields = {};
+    Object.entries(sectionJson).map(([title, object]) => {
+        // We only care about children that are cards/Answers or cards:AnswerSections
+        if (object["sling:resourceSuperType"] == "cards/Answer" && object.statusFlags?.includes("INCOMPLETE")) {
```
It might be wiser to check for the `DRAFT` flag which should cover both INCOMPLETE and INVALID answers (e.g. if someone enters -15 as their year of birth the answer should be flagged as INVALID)

veronikaslc replied:
Addressed in [6f05307](https://github.com/data-team-uhn/cards/pull/1282/commits/6f05307b3febe8028d156fb75bb3167deda6044d)

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Separated utility functions to a separate utility file

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Added license header to utility file

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtility.jsx:
```diff
@@ -0,0 +1,86 @@
+    return retFields;
+}
+
+  // Recursively collect all uuids of questions in the *right order*
+export function parseSectionOrQuestionnaire (sectionJson) {
```
```suggestion
// Recursively collect all uuids of questions in the *right order*
export function parseSectionOrQuestionnaire (sectionJson) {
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtility.jsx:
```diff
@@ -0,0 +1,86 @@
+    return retFields;
+}
+
+export function getFirstIncompleteQuestionEl (json) {
```
Please add comment

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtility.jsx:
```diff
@@ -0,0 +1,86 @@
+// A helper recursive function to loop through the sections/questions data 
+// and collect all questions that have incomplete status among it's statusFlags
+// Recursively build a map of uuid->path of all incomplete questions
+export function getIncompleteQuestionsMap (sectionJson) {
```
I'm not sure the first 2 functions need to be exported as they seem a bit technical, but I guess there's no real harm in exposing them.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtility.jsx:
```diff
@@ -0,0 +1,86 @@
+
+    }
+    return null;
+}
```
Please add newline at the end of file per our code style guidelines.

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Added comments to the utility function

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Better codestyle

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
* Renamed the new FormUtility to FormUtilities
* Moved the utility function `hasWarningFlags` from `AnswerInstructions` to FormUtilities
* Modified hasWarningFlags
  * to accept either a [UUID, dataObject] pair or directly a data object as a parameter
  * to also return true for the DRAFT flag
* Modified the criteria for requireCompletion to use hasWarningFlags instead of looking only for the INCOMPLETE flag

marta- commented:
The changes in the last commit include some minor code refactoring. Regression testing needed:
* the color of the answer instructions (e.g. "This question is mandatory") correctly reflects the presence of the answer in all questions including QuestionMatrix
* Incomplete questions and sections appear in view mode, while non-mandatory questions without an answer are not displayed in view mode

Also rebased on the latest `dev`.

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Set requireCompletion to True in the Rehab survey

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Made the red outline of missing required fields lighter

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -410,7 +456,21 @@ function Form (props) {
+	                         incompleteQuestionEl?.classList.remove(classes.questionnaireItemWithError);
+	                         setIncompleteQuestionEl(null);
+	                         paginationEnabled && setDisableProgress(requireCompletion);
+	                         setIncompleteQuestionEl(null);
```
Duplicated call to `setIncompleteQuestionE1`

veronikaslc replied:
@marta- Seems harmless to address here, wdyt?

marta- replied:
Removed the duplicate in 80d8cf4.

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Addressing review comments: remove duplicate line

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Problems found in PROMS mode
- set `Require completion: Yes` on the SF-12 form
- edit the form, leave some fields empty
- save; an error appears, however the save action gets completed

- log in as patient
- do not complete the SF12 forms
- move to the next survey
- this should not be allowed

sashaandjic commented:
Problem in PREMS mode.
Remove the pagination from any questionnaire.
Edit the form (either as admin or as a patient).
Saving should be prevented as every `prems` form has `Require completion: Yes`
However, you can save incomplete forms as an admin, and move to the next survey as a patient.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -222,6 +254,20 @@ function Form (props) {
+                    setDisableProgress(false);
+                  }
+              })
+              .catch(handleFetchError);
```
Remove lines 248-249 and:

```suggestion
              .catch(handleFetchError)
              .finally(() => {
                setLastSaveStatus(true);
                setLastSaveTimestamp(new Date());
              });
        } else {
          setLastSaveStatus(true);
          setLastSaveTimestamp(new Date());
```

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Block progressing on the non-paginated forms

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Fixed whitespacing

veronikaslc committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Fixed matrix section error

acrowthe commented:
New commits look fine

marta- reviewed (approved):
The last fixes look right

sashaandjic commented:
For some starnge reasons I can leave this pages of CPES survey incomplete:
![image](https://user-images.githubusercontent.com/23243244/215225625-69015d94-ad63-4e68-b724-11c4f7b33000.png)

I did not see anything like this on ED and IP surveys.
I was playing with yes/no question on this page, where each brings different set of fields. Every time I switched the state of yes/no the new set of empty questions would appear. For some strange reason i can leave them empty and proceed to the next page.

marta- commented:
> For some starnge reasons I can leave this pages of CPES survey incomplete

Note that the 2nd question is not mandatory. If we make it mandatory, i think there are some other bugs that pop up because it's a conditional section. So it's not a bug with 1375, it's by design.

marta- commented:
> > For some starnge reasons I can leave this pages of CPES survey incomplete
> 
> Note that the 2nd question is not mandatory. If we make it mandatory, i think there are some other bugs that pop up because it's a conditional section. So it's not a bug with 1375, it's by design.

I tested with the questions inside the conditional sections modified to be mandatory, and the behavior is correct.

marta- commented:
> > > For some starnge reasons I can leave this pages of CPES survey incomplete
> > 
> > 
> > Note that the 2nd question is not mandatory. If we make it mandatory, i think there are some other bugs that pop up because it's a conditional section. So it's not a bug with 1375, it's by design.
> 
> I tested with the questions inside the conditional sections modified to be mandatory, and the behavior is correct.

I also can't find anything wrong happening in this particular questionnaire if the questions inside conditional sections are mandatory, so I may make a separate task and PR to change them after some more testing.

sashaandjic commented:
`Require completion` functionality still does not work on non-paginated forms. Steps to reproduce:
- start cards in `prems` mode
- go to the designer and remove pagination from any survey
- create a visit info form for the modified survey form to be created
- login as patient
- as soon as you land on the first page click on the `REVIEW` button
- as expected you will be redirected to the very first empty question of the form. The question will be "circled" with the red box
- scroll down and notice that the `REVIEW` button turned into `SAVED` button
- click on the `SAVED` button and you will be able to proceed past the survey
![image](https://user-images.githubusercontent.com/23243244/215230360-1d23f1bf-2dd7-4ec4-a225-25a0f971ed90.png)
![image](https://user-images.githubusercontent.com/23243244/215230393-2c9bdcd6-66d0-4319-961a-0c8fbd381444.png)

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Set `disableProgress` to true right before the form re-fetch that determines whether all the mandatory questions on the screen have been answered.
Fixes bug where you can still advance from incomplete non-paginated forms in the patient UI.

marta- committed:
[misc] Don't show "Saved" when we have a custom label for the form's save button

marta- commented:
> `Require completion` functionality still does not work on non-paginated forms. Steps to reproduce:

Should be fixed by 6aff5e1.

Also "Saved" should no longer appear in the patient ui (only in the admin ui) since 37c45b6.

marta- committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
When completion is required, disable progress after the form has changed only if is paginated.
On non-paginated forms, ensure progress is enabled until we launch the call to check the updated flags.

sashaandjic commented:
on non-paginated forms the disabled REVIEW bttn does not become enabled once I fill in all required fields. the only way to proceed to the other page (either other survey or review page) is by doing CTRL+R

sashaandjic reviewed (approved):
tested - `proms` & `prems` **all good**

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -71,6 +71,7 @@ function Question (props) {
 
   return (
     <Card
+      id={questionDefinition["@path"]}
```
This will cause problems with repeatable sections. Can't the actual UUID of the node be used? Even if there's no UUID for answers now, it's easy to add one.

marta- replied:
Good catch...

sdumitriu replied:
Since there are no repeatable sections in prems, I think we can merge this as is and make a task to switch to UUID in the (near) future.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtilities.jsx:
```diff
@@ -0,0 +1,97 @@
+          // If this is an cards:Question, copy the entire thing over to our Json value
+          retFields.push(object["jcr:uuid"]);
+        } else if (object["jcr:primaryType"] == "cards:Section") {
+          if ("matrix" === object["displayMode"]) {
```
So this highlights the whole matrix? Can we go deeper and highlight just a row from the matrix?

marta- replied:
I think this is ok. The message "This question is mandatory" is actually under the matrix title. Users tend to see the matrix as a whole anyway,

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormUtilities.jsx:
```diff
@@ -0,0 +1,97 @@
+      let incompleteQuestions = getIncompleteQuestionsMap(json);
+
+      if (Object.keys(incompleteQuestions).length > 0) {
+	// Loop through the question uuids in the order defined by questionnaire
```
After CARDS-2054, looking at the questionnaire order won't be necessary anymore.

marta- replied:
Let's make a note on 2054 to simplify this.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -561,6 +561,11 @@ const questionnaireStyle = theme => ({
     },
+    questionnaireItemWithError: {
+      "&.MuiCard-root, > .MuiCard-root" : {
+        outline: `1px solid ${theme.palette.error.light}`,
```
Why not `2px` like the focused class above?

marta- replied:
I don't like 2px in red, too thick and accusing. 1px looks best in my opinion.

sdumitriu replied:
I prefer 2px, but it's up to you.

sdumitriu reviewed (commented):
Can we add the new property in the questionnaire nodetype definition in `/modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd`?

sdumitriu committed:
CARDS-1375: Add an option to block advancing in a form unless all required fields have been filled out
Add the new property to the CND

sdumitriu committed:
[misc] Replace spaces with tabs, to match the rest of the file
